core/src/test/java/org/elasticsearch/index/query/SimpleQueryStringBuilderTest.java

```diff
@@ -0,0 +1,384 @@
+
+        qb.defaultOperator(Operator.AND);
+        boolQuery = (BooleanQuery) qb.toQuery(createContext());
+        assertThat(shouldClauses(boolQuery), is(0));
+
+        qb.defaultOperator(Operator.OR);
+        boolQuery = (BooleanQuery) qb.toQuery(createContext());
+        assertThat(shouldClauses(boolQuery), is(4));
+    }
+
+    @Test
+    public void testValidation() {
+        SimpleQueryStringBuilder qb = createTestQueryBuilder();
+        assertNull(qb.validate());
+    }
+
+    @Test
+    public void testNullQueryTextGeneratesException() {
+        SimpleQueryStringBuilder builder = new SimpleQueryStringBuilder(null);
+        QueryValidationException exception = builder.validate();
+        assertThat(exception, notNullValue());
+    }
+
+    @Test
+    public void testHandlingDefaults() throws IOException {
+        SimpleQueryStringBuilder qb = createTestQueryBuilder();
+        qb.analyzer(null);
+        qb.minimumShouldMatch(null);
+        qb.queryName(null);
+        assertEquals(qb.toQuery(createContext()), createExpectedQuery(qb, createContext()));
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testFieldCannotBeNull() {
+        SimpleQueryStringBuilder qb = createTestQueryBuilder();
+        qb.field(null);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testFieldCannotBeNullAndWeighted() {
+        SimpleQueryStringBuilder qb = createTestQueryBuilder();
+        qb.field(null, 1.0f);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testFieldCannotBeEmpty() {
+        SimpleQueryStringBuilder qb = createTestQueryBuilder();
+        qb.field("");
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testFieldCannotBeEmptyAndWeighted() {
+        SimpleQueryStringBuilder qb = createTestQueryBuilder();
+        qb.field("", 1.0f);
+    }
+
+    /**
+     * The following should fail fast - never silently set the map containing
+     * fields and weights to null but refuse to accept null instead.
+     * */
+    @Test(expected = NullPointerException.class)
+    public void testFieldsCannotBeSetToNull() {
+        SimpleQueryStringBuilder qb = createTestQueryBuilder();
+        qb.fields(null);
+    }
+
+    /*
+     * This assumes that Lucene query parsing is being checked already, adding
+     * checks only for our parsing extensions.
+     * 
+     * Also this relies on {@link SimpleQueryStringTests} to test most of the
+     * actual functionality of query parsing.
+     */
+    @Override
+    protected void assertLuceneQuery(SimpleQueryStringBuilder queryBuilder, Query query, QueryParseContext context) {
+        if (queryBuilder.queryName() != null) {
+            Query namedQuery = context.copyNamedFilters().get(queryBuilder.queryName());
+            assertThat(namedQuery, equalTo(query));
+        }
+
+        if (queryBuilder.fields().size() > 1 && Strings.hasText(queryBuilder.text())) {
```

Comment:
I missed your rely to my previous comment. Answer inline:

> not sure I follow, how are these checks independent of the query builder to 
> lucene query conversion code? You mean indepedent of createExpectedQuery 
> code?

At least in this case createExpectedQuery in the test is almost identical to the toQuery code in the class under test.

The code in assertLuceneQuery is independent in that it's checks don't rely on building the correct Lucene query to do the check.

> once through createExpectedQuery and then comparing the result with 
> queryBuilder.toQuery, once in the assertLuceneQuery. Isn't this the case?

We are testing the same things twice. I saw only now that assertLuceneQuery runs after the equality comparison with createExpectedQuery. I had expected this to happen the other way around. Then one of the arguments would have been that assertLuceneQuery should be less likely to break if there are little optimisations in which Lucene query is generated. 

This way the only weak argument in favour of the implementation that remains is remembering that tests also serve a documentation purpose, where this method shows what type of Lucene query to expect given some input. Not sure if this is valuable information to have though.